aws_protocol_adaptor/device_client/aws_nvmsgbroker.c

```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
 	last_send_time_stamp = current_time_stamp;
 	return;
 }
+
+char *nvds_msgapi_getversion()
+{
+	return (char *)NVDS_MSGAPI_VERSION;
+}
+
+char *nvds_msgapi_get_protocol_name()
+{
+	return (char *)NVDS_MSGAPI_PROTOCOL;
+}
+
+bool is_valid_connection_str(char *connection_str, char *&burl, char *&bport)
+{
+	if (connection_str == NULL)
+	{
+		IOT_ERROR("connection string cant be NULL");
+		return false;
+	}
+
+	char conn_str[] = connection_str;
+	int i = 0;
+
+	char *token = strtok(conn_str, ";");
+	char *data[2];
+
+	while (token)
+	{
+		data[i++] = token;
+		token = strtok(NULL, ";");
+	}
+	for (i = 0; i < 2; i++)
+	{
+		printf("%s\n", data[i]);
+	}
+	char *burl = data[0];
+	char *bport = data[1];
+
+	if (burl == "" || bport == "")
+	{
+		IOT_ERROR("connection string is invalid. hostname or port is empty\n");
+		return false;
+	}
+	return true;
+}
+
+char *generate_sha256_hash(char *str)
+{
+	unsigned char hashval[SHA256_DIGEST_LENGTH];
+	int len = SHA256_DIGEST_LENGTH * 2 + 1;
+	char res[len];
```

Comment:
`res` is a statically allocated variable, but you're trying to return it. What do you think will happen as a result?

Reply:
To remedy the problem, you should supply the output string as an argument to the function. Note: it will need to have space for at least 65 characters (which you'll need to add a check for).

Reply:
The max_len is actually the length of the output_str and I have added a check in the connection signature to check if has at least 65 characters

Reply:
Sounds good, but you still need to resolve the issue of returning a statically allocated variable. res is created on the stack, not the heap, so its lifetime is only that of the enclosing scope. The value of res will be invalidated once this function returns.